src/onnx/parse_scan.cpp

```diff
@@ -0,0 +1,314 @@
+        const auto m   = info.attributes["num_scan_inputs"].i();
+        const auto n   = args.size() - m;
+        const auto k   = body_outs.size() - n;
+
+        std::vector<instruction_ref> body_params;
+        transform(body->get_parameter_names(),
+                  std::back_inserter(body_params),
+                  [&](const auto& name) { return body->get_parameter(name); });
+
+        if(auto num_body_ins = body_params.size(); num_body_ins != n + m)
+            MIGRAPHX_THROW("Scan: Number of inputs to body {" + std::to_string(num_body_ins) +
+                           "} does not match number of inputs to Scan {" + std::to_string(n + m) +
+                           "}");
+
+        const auto scan_input_axes = parse_axes(info, "scan_input_axes", m, args.begin() + n, 0);
+        const auto scan_input_directions = parse_dirs(info, "scan_input_directions", m);
+        const auto scan_output_axes =
+            parse_axes(info, "scan_output_axes", k, body_outs.begin() + n, 1);
+        const auto scan_output_directions = parse_dirs(info, "scan_output_directions", k);
+
+        // Check that scan axes lens are the same across all scan inputs
+        size_t num_iters = args[n]->get_shape().lens()[scan_input_axes[0]];
+        for(auto i = 1; i < m; ++i)
+            if(args[n + i]->get_shape().lens()[scan_input_axes[i]] != num_iters)
+                MIGRAPHX_THROW(
+                    "Scan: Lengths of scan_input_axes do not match across all scan inputs.\n"
+                    "Scan input shapes: " +
+                    to_string_range(
+                        to_shapes(std::vector<instruction_ref>(args.begin() + n, args.end()))) +
+                    "\nScan input axes: " + to_string_range(scan_input_axes));
+
+        if(num_iters > parser.max_loop_iterations)
+            MIGRAPHX_THROW("Scan: Number of required iterations {" + std::to_string(num_iters) +
+                           "} would exceed the maximum iteration limit {" +
+                           std::to_string(parser.max_loop_iterations) + "}");
+
+        // Check that state variable shapes match between the Scan node and its body attribute
+        for(auto i = 0; i < n; ++i)
+            if(args[i]->get_shape() != body_params[i]->get_shape())
+                MIGRAPHX_THROW("Scan: State input " + std::to_string(i) + " shape {" +
+                               to_string(args[i]->get_shape()) +
+                               "} does not match corresponding body input shape {" +
+                               to_string(body_params[i]->get_shape()) + "}");
+
+        // Check that the shapes of scan inputs sliced across scan input axes match the shapes of
+        // the body attribute scan inputs
+        for(auto i = 0; i < m; ++i)
+        {
+            auto node_shape = args[i + n]->get_shape();
+            auto node_lens  = node_shape.lens();
+            node_lens.erase(node_lens.begin() + scan_input_axes[i]);
```

Comment:
Why is it erased ? 
I thought slice shape would make dimension on slice axis 1. But other dimensions would remain the same.

Reply:
The ONNX spec defines it that way:
["Note that the iterated element passed to the body subgraph does not have a sequence axis. It will have a rank one less than the rank of the corresponding scan_input"](https://onnx.ai/onnx/operators/onnx__Scan.html#scan-21).